examples/vision/pointnet.py

```diff
@@ -0,0 +1,283 @@
+            test_labels.append(i)
+
+    return (
+        np.array(train_points),
+        np.array(test_points),
+        np.array(train_labels),
+        np.array(test_labels),
+        class_map,
+    )
+
+"""
+Set the number of points to sample and batch size and parse the dataset. This can take
+~5minutes to complete.
+"""
+
+NUM_POINTS = 2048
+NUM_CLASSES = 10
+BATCH_SIZE = 32
+
+train_points, test_points, train_labels, test_labels, CLASS_MAP = parse_dataset(
+    NUM_POINTS
+)
+
+"""
+Our data can now be read into a `tf.data.Dataset()` object. We set the shuffle buffer
+size to the entire size of the dataset as prior to this the data is ordered by class.
+Data augmentation is important when working with point cloud data. We create a
+augmentation function to jitter and shuffle the train dataset.
+"""
+
+def augment(points, label):
+    # jitter points
+    points += tf.random.uniform(points.shape, -0.005, 0.005, dtype=tf.float64)
+    # shuffle points
+    points = tf.random.shuffle(points)
+    return points, label
+
+train_dataset = tf.data.Dataset.from_tensor_slices(
+    (train_points, train_labels))
+test_dataset = tf.data.Dataset.from_tensor_slices((test_points, test_labels))
+
+train_dataset = train_dataset.shuffle(
+    len(train_points)).map(augment).batch(BATCH_SIZE)
+test_dataset = test_dataset.shuffle(len(test_points)).batch(BATCH_SIZE)
+
+"""
+### Build a model
+
+Each convolution and fully-connected layer (with exception for end layers) consits of
+Convolution / Dense -> Batch Normalization -> ReLU Activation.
+"""
+
+def conv_bn(x, filters):
+    x = layers.Conv1D(filters, kernel_size=1, padding="valid")(x)
+    x = layers.BatchNormalization(momentum=0.0)(x)
+    return layers.Activation("relu")(x)
+
+def dense_bn(x, filters):
+    x = layers.Dense(filters)(x)
+    x = layers.BatchNormalization(momentum=0.0)(x)
+    return layers.Activation("relu")(x)
+
+"""
+PointNet consists of two core components. The primary MLP network, and the transformer
+net (T-net). The T-net aims to learn an affine transformation matrix by its own mini
+network. The T-net is used twice. The first time to transform the input features (n, 3)
+into a canonical representation. The second is an affine transformation for alignment in
+feature space (n, 3). As per the original paper we constrain the transformation to be
+close to an orthogonal matrix (i.e. ||X*X^T - I|| = 0).
+"""
+
+class OrthogonalRegularizer(keras.regularizers.Regularizer):
+    def __init__(self, k, l2reg=0.001):
+        self.k = k
+        self.l2reg = l2reg
+        self.eye = tf.eye(self.k)
+
+    def __call__(self, x):
+        x = tf.reshape(x, (-1, self.k, self.k))
+        xxt = tf.tensordot(x, x, axes=(2, 2))
+        xxt = tf.reshape(xxt, (-1, self.k, self.k))
+        return tf.reduce_sum(self.l2reg * tf.square(xxt - self.eye))
+
+"""
+ We can define the T-net as a functional model.
+"""
+
+def tnet(num_points, k, name):
```

Comment:
Please use a more descriptive name for `k`.

Going back to the comment about using a functional approach -- return a `Model` instance is fine, but in this case it turns out you don't need to do it, because you never use that instance as a model (e.g. `fit()`). So here converting this function to just apply layers to an input (like `conv_bn`) would be beneficial, since it would mean you don't need to pass `num_points`.

Reply:
k is the number of features in the final dimension i.e. (batch, num_points, k). Could change to num_feats, or num_channels?

Reply:
`num_features` sounds good to me, clear & descriptive.